ethereum/event/BREvent.c

```diff
@@ -168,48 +142,45 @@ eventHandlerAlarmCallback (BREventHandler handler,
     eventHandlerSignalEventOOB (handler, (BREvent*) &event);
 }
 
-
 typedef void* (*ThreadRoutine) (void*);
 
 static void *
 eventHandlerThread (BREventHandler handler) {
-
 #if defined (__ANDROID__)
     pthread_setname_np(handler->thread, handler->name);
 #else
     pthread_setname_np(handler->name);
 #endif
 
-    pthread_mutex_lock (handler->lockToUse);
+    int timeToQuit = 0;
 
-    handler->threadQuit = 0;
-
-    while (!handler->threadQuit) {
+    while (!timeToQuit) {
         // Check for a queued event
-        switch (eventQueueDequeue(handler->queue, handler->scratch)) {
+        switch (eventQueueDequeueWait (handler->queue, handler->scratch)) {
             case EVENT_STATUS_SUCCESS: {
                 // If we have one, dispatch
+                if (handler->lockOnDispatch) pthread_mutex_lock (handler->lockOnDispatch);
                 BREventType *type = handler->scratch->type;
                 type->eventDispatcher (handler, handler->scratch);
-                break;
+                if (handler->lockOnDispatch) pthread_mutex_unlock (handler->lockOnDispatch);
+               break;
             }
 
+            case EVENT_STATUS_WAIT_ABORT:
+                timeToQuit = 1;
+                break;
+
             case EVENT_STATUS_NOT_STARTED:
             case EVENT_STATUS_UNKNOWN_TYPE:
             case EVENT_STATUS_NULL_EVENT:
+            case EVENT_STATUS_WAIT_ERROR:
```

Comment:
Just based on the "impossible?" comment, do we want to assert on (some of?) these, to see if they occur?

Reply:
Fixed.